### PR TITLE
Improve retry logic

### PR DESCRIPTION
**Pull request by ethan-tonic**
This PR adds retry logic for when scoring fails due to parsing issues. It also adds exponential retries for openai.

### CODE REVIEW

File: tonic_validate/validate_scorer.py [hunk 1]
```diff
@@ -40,11 +44,17 @@ def __init__(
...
+        max_parsing_retries: int
+            The number of times to retry a failed score if it failed due to parsing.
+        max_llm_retries: int
+            The number of times to retry a failed llm request.
         fail_on_error: bool
             If True, an error in calculating a metric will raise an exception. If False, the score will be set to None.
         """
         self.metrics = metrics
         self.model_evaluator = model_evaluator
+        self.max_scoring_retries = max_parsing_retries
```
Comment (joeferraratonic) [inline tonic_validate/validate_scorer.py:56]: Why isn't the naming consistent?

Comment (joeferraratonic) [reply tonic_validate/validate_scorer.py:56]: This variable, `self.max_scoring_retries` is never used, so it along with the parameter `max_parsing_retries` should be removed.

File: tonic_validate/validate_scorer.py [hunk 2]
```diff
@@ -56,20 +66,44 @@ def __init__(
...
-                score = metric.score(response, openai_service)
-            except Exception as e:
-                if not self.fail_on_error:
-                    score = None
-                    logger.error(
-                        f"Error calculating score for {metric.name}, setting score to None."
+            tries = 0
+            exceptions = []
+            last_cache = copy.deepcopy(openai_service.cache)
+            while tries < 3:
```
Comment (joeferraratonic) [inline tonic_validate/validate_scorer.py:76]: Should this be `while tries < self.max_scoring_retries`?

File: tonic_validate/validate_scorer.py (hunk 1 above)
Comment (joeferraratonic) [reply tonic_validate/validate_scorer.py:56]: See the comment below. I think you forgot to use this variable.

**Review by joeferraratonic — CHANGES_REQUESTED** [review]
Aside form my comments, this looks good.

File: tonic_validate/validate_scorer.py (hunk 2 above)
Comment (ethan-tonic) [reply tonic_validate/validate_scorer.py:76]: I think I accidentally left that in. My bad

**Review by joeferraratonic — APPROVED** [review]
Thanks for the changes, looks good now.

**Pull request merged**